### PR TITLE
Producing to single feedback topic in event of failures

### PR DESCRIPTION
**Pull request by laurence-bird**

### CODE REVIEW

File: src/main/scala/com/ovoenergy/comms/composer/kafka/EventProcessor.scala [hunk 1]
```diff
@@ -28,54 +23,42 @@ object EventProcessor extends Logging {
...
-    def sendOutput(event: OutEvent): F[RecordMetadata] = {
-      outputProducer(event).onError {
-        case NonFatal(e) =>
-          Async[F].delay(
-            warnWithException(event)(
-              "Unable to produce event, however, processing has completed so offset will be committed regardless")(e))
-      }
-    }
+      implicit buildFeedbackFrom: BuildFeedback[InEvent],
+      f: Effect[F]): Record[InEvent] => F[Seq[RecordMetadata]] = { record: Record[InEvent] =>
```
Comment (filosganga) [inline src/main/scala/com/ovoenergy/comms/composer/kafka/EventProcessor.scala]: I believe the convention is to have `f` uppercase as `F`.

File: src/main/scala/com/ovoenergy/comms/composer/kafka/EventProcessor.scala [hunk 2]
```diff
@@ -28,54 +23,42 @@ object EventProcessor extends Logging {
...
+      } yield Seq(r1, r2)
+    }
 
-      failedProducer(failed).onError {
-        case NonFatal(e) =>
-          Async[F].delay(warnWithException(failed)(
-            "Unable to produce Failed event, however, processing has completed so offset will be committed regardless")(
-            e))
+    def handleEvent(inEvent: InEvent) = {
+      info(inEvent)("Processing event")
```
Comment (filosganga) [inline src/main/scala/com/ovoenergy/comms/composer/kafka/EventProcessor.scala]: I would wrap it in an `F`

File: src/test/scala/com/ovoenergy/comms/composer/kafka/EventProcessorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    val successfulProcessEvent = (input: InputEvent) => Right(OutputEvent(generate[String]))
+    val eventProcessor: Record[InputEvent] => IO[Seq[RecordMetadata]] =
+      EventProcessor[IO, InputEvent, OutputEvent](successfulOutputProducer.function,
+                                                  successfulFailedProducer.function,
+                                                  successfulFeedbackProducer.function,
+                                                  successfulProcessEvent)
+
+    val record = new ConsumerRecord[Unit, Option[InputEvent]]("input", 1, 1, (), Some(inputEvent))
+
+    eventProcessor(record).unsafeRunSync() should contain theSameElementsAs Seq(successfulOutputProducer.result)
```
Comment (filosganga) [inline src/test/scala/com/ovoenergy/comms/composer/kafka/EventProcessorSpec.scala]: you can use contain only as well

File: src/main/scala/com/ovoenergy/comms/composer/kafka/EventProcessor.scala [hunk 4]
```diff
@@ -28,54 +23,43 @@ object EventProcessor extends Logging {
...
-          Async[F].delay(warn(record)(s"Failed to deserialise kafka record ${record.show}"))
+          F.delay(warn(record)(s"Failed to deserialise kafka record ${record.show}"))
+            .map(_ => Seq.empty[RecordMetadata])
         }
-      })
-    }
-
-    result
+      }
+    } yield res
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/composer/kafka/EventProcessor.scala]: I find the style you have used above with `>>` and `match` a bit clearer, personally.
`map(_ => a)` is `as(a)` (this is just to make you aware, you don't have to change it).

**Comment by SystemFw** [comment]
There's a compilation error in the build.sbt as well

**Comment by ovo-comms-circleci** [comment]
This PR was successfully deployed to production by [CircleCI build 365](https://circleci.com/gh/ovotech/comms-composer/365) at 2018-08-20 16:37 BST.

**Pull request merged**